apps/dialer/js/handled_call.js

```diff
@@ -0,0 +1,474 @@
+    } else {
+      FontSizeManager.resetFixedBaseline(this.numberNode);
+    }
+};
+
+/**
+ * This method renders a phone number. It can be used to display numbers typed
+ * during a call without altering the original number stored in the object.
+ */
+HandledCall.prototype.renderPhoneNumber =
+  function hc_renderPhoneNumber(number, ellipsisSide) {
+    if (typeof(number) === 'string') {
+      navigator.mozL10n.setAttributes(this.numberNode, number);
+    } else if (number.id) {
+      navigator.mozL10n.setAttributes(this.numberNode, number.id,
+                                      number.args);
+      this.numberNode.style.fontSize = '';
+    } else if (number.raw) {
+      this.numberNode.removeAttribute('data-l10n-id');
+      this.numberNode.textContent = number.raw;
+      this.formatPhoneNumber(ellipsisSide);
+    }
+};
+
+/**
+ * This method replaces the phone number in the call with a new number and
+ * renders it.
+ */
+HandledCall.prototype.replacePhoneNumber =
+  function hc_replacePhoneNumber(number, ellipsisSide) {
+    this.renderPhoneNumber(number, ellipsisSide);
+    this._cachedInfo = number;
+};
+
+/**
+ * This method renders the number that was originally associated with this
+ * object in case it was altered (e.g. by typing numbers during a call).
+ */
+HandledCall.prototype.restorePhoneNumber =
+  function hc_restorePhoneNumber() {
+    this._cachedInfo && this.replacePhoneNumber(this._cachedInfo, 'end');
+};
+
+HandledCall.prototype.updateDirection = function hc_updateDirection() {
+  var classList = this.node.classList;
+  if (this._initialState === 'incoming') {
+    classList.add('incoming');
+    this.node.setAttribute('data-l10n-id', 'incoming_screen_reader');
+  } else {
+    classList.add('outgoing');
+    this.node.setAttribute('data-l10n-id', 'outgoing_screen_reader');
+  }
+
+  if (this.call.state === 'connected') {
+    classList.add('ongoing');
+  }
+};
+
+HandledCall.prototype.remove = function hc_remove() {
+  this._removed = true;
+  this.call.removeEventListener('statechange', this);
+  this.photo = null;
+
+  var self = this;
+  CallScreen.stopTicker(this.durationNode);
+  var currentDuration = ConferenceGroupHandler.isGroupDetailsShown() ?
+    ConferenceGroupHandler.currentDuration : this.durationChildNode.textContent;
+
+  this.node.classList.add('ended');
+  this.durationNode.classList.remove('isTimer');
+
+  /* This string will be resized automatically to fit its container, see
+   * hc_observeMutations() & hc_computeCallEndedFontSizeRules. */
+  navigator.mozL10n.setAttributes(this.durationChildNode, 'callEnded');
+
+  // FIXME/bug 1007148: Refactor duration element structure. No number or ':'
+  // existence checking will be necessary.
+  var totalDuration = !!currentDuration.match(/\d+/g) ? currentDuration : '';
+  this.totalDurationNode.textContent = totalDuration;
+
+  setTimeout(function(evt) {
+    CallScreen.removeCall(self.node);
+    CallsHandler.end();
```

Comment:
Why did you add the `end()` call here? In theory `hc_remove()` should be called only when a call has already ended so I think this is redundant.